### PR TITLE
Release 2.17.0-rc.0

### PR DESCRIPTION
**Pull request by roidelapluie**
Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,25 @@
+## 2.17.0-rc.0 / 2020-03-11
+
+This release implements isolation in TSDB. API queries and recording rules are guaranteed to only see full scrapes and full recording rules. This causes a slight increase in memory usage.
+
+* [ENHANCEMENT] PromQL: Allow more keywords as metric names #6933
+* [ENHANCEMENT] React UI: Adds normalization of localhost urls in targets page react #6794
```
Comment (roidelapluie) [inline CHANGELOG.md]:
```suggestion
* [ENHANCEMENT] React UI: Adds normalization of localhost urls in targets page #6794
```

**Comment by simonpasquier** [comment]
LGTM

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,24 @@
+## 2.17.0-rc.0 / 2020-03-11
+
+This release implements isolation in TSDB. API queries and recording rules are guaranteed to only see full scrapes and full recording rules. This causes a slight increase in memory usage.
```
Comment (beorn7) [inline CHANGELOG.md]: "…see full scrapes and full recording rules." → "…see the full result of a scrape or rule evaluation."

"This causes a slight increase in memory usage." → It's not only memory usage. Perhaps something like: "This comes with a certain overhead in resource usage. Depending on the situation, there might be a noticeable increase in memory usage, CPU usage, or query latency."

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,24 @@
...
+
+* [ENHANCEMENT] PromQL: Allow more keywords as metric names #6933
+* [ENHANCEMENT] React UI: Add normalization of localhost urls in targets page #6794
```
Comment (beorn7) [inline CHANGELOG.md]: urls → URLs

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,24 @@
...
+* [ENHANCEMENT] Rules: Make deleted rule series as stale after a reload #6745
```
Comment (beorn7) [inline CHANGELOG.md]: Make → Mark

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,24 @@
...
+* [ENHANCEMENT] Scrape: Log scrape append failures as debug rather than warn #6852
+* [ENHANCEMENT] TSDB: Support isolation #6841
+* [ENHANCEMENT] TSDB: Log WAL replay duration #6922
+* [ENHANCEMENT] TSDB: Improve query performance for queries that partially hit the head #6676
+* [ENHANCEMENT] Consul SD: Add health label to metrics #5313
+* [ENHANCEMENT] EC2 SD: Expose EC2 instance lifecycle as label #6914
+* [ENHANCEMENT] Kubernetes SD: add service type metadata to k8s service role #6496 #6684
```
Comment (beorn7) [inline CHANGELOG.md]: add → Add
k8s → K8s

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,24 @@
...
+* [ENHANCEMENT] Kubernetes SD: expose label_selector and field_selector #6807
```
Comment (beorn7) [inline CHANGELOG.md]: expose → Expose

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,24 @@
...
+* [ENHANCEMENT] Openstack SD: Add openstack_hypervisor_id meta label #6962
+* [BUGFIX] PromQL: Do not escape html-like chars in query log #6834 #6795
```
Comment (beorn7) [inline CHANGELOG.md]: html → HTML

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,24 @@
...
+* [BUGFIX] React UI: Fix data table matrix values #6896
+* [BUGFIX] React UI: Fix new targets page not loading when using non-ascii characters #6892
```
Comment (beorn7) [inline CHANGELOG.md]: ascii → ASCII

File: CHANGELOG.md (hunk 2 above)
Comment (roidelapluie) [reply CHANGELOG.md]: Indeed. Updated.

File: CHANGELOG.md (hunk 3 above)
Comment (roidelapluie) [reply CHANGELOG.md]: Thanks

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,27 @@
+## 2.17.0-rc.0 / 2020-03-11
+
+This release implements isolation in TSDB. API queries and recording rules are
+guaranteed to only see full scrapes and full recording rules. This comes with a
+certain overhead in resource usage. Depending on the situation, there might be a
```
Comment (bwplotka) [inline CHANGELOG.md]: It sounds like "Inevitable you will see certain overhead". It really IMO sounds more negative then it is... I think we can tell this softer.

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,27 @@
...
+noticeable increase in memory usage, CPU usage, or query latency.
```
Comment (bwplotka) [inline CHANGELOG.md]:
```suggestion
some increase in memory usage, CPU usage, or query latency.
```

Comment (roidelapluie) [reply CHANGELOG.md]: @beorn7 WDYT?

Comment (beorn7) [reply CHANGELOG.md]: I'm a fan of "underpromise, overdeliver", and I found my wording already quite cautious (with "might" and such), but I don't feel strongly.

But if you do want to change "noticeable" to "some", then you have to delete the "a" in the previous line.

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,27 @@
...
+
+* [ENHANCEMENT] PromQL: Allow more keywords as metric names #6933
+* [ENHANCEMENT] React UI: Add normalization of localhost URLs in targets page #6794
+* [ENHANCEMENT] Rules: Mark deleted rule series as stale after a reload #6745
+* [ENHANCEMENT] Scrape: Log scrape append failures as debug rather than warn #6852
+* [ENHANCEMENT] TSDB: Support isolation #6841
```
Comment (brian-brazil) [inline CHANGELOG.md]: This is a FEATURE, it's pretty big

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,3 +1,27 @@
...
+* [ENHANCEMENT] TSDB: Log WAL replay duration #6922
+* [ENHANCEMENT] TSDB: Improve query performance for queries that partially hit the head #6676
+* [ENHANCEMENT] Consul SD: Add health label to metrics #5313
```
Comment (brian-brazil) [inline CHANGELOG.md]: To metadata?

File: CHANGELOG.md [hunk 13]
```diff
@@ -1,3 +1,27 @@
...
+guaranteed to only see full scrapes and full recording rules. This comes with a
+certain overhead in resource usage. Depending on the situation, there might be a
+noticeable increase in memory usage, CPU usage, or query latency.
+
+* [ENHANCEMENT] PromQL: Allow more keywords as metric names #6933
+* [ENHANCEMENT] React UI: Add normalization of localhost URLs in targets page #6794
+* [ENHANCEMENT] Rules: Mark deleted rule series as stale after a reload #6745
+* [ENHANCEMENT] Scrape: Log scrape append failures as debug rather than warn #6852
+* [ENHANCEMENT] TSDB: Support isolation #6841
+* [ENHANCEMENT] TSDB: Log WAL replay duration #6922
```
Comment (brian-brazil) [inline CHANGELOG.md]: This seems a bit small to mention

**Comment by roidelapluie** [comment]
Alright, I merge & publish the RC tomorrow morning.

**Comment by roidelapluie** [comment]
@beorn7 Instantiating a queries now takes a very long time:
prom

Maybe that's worth taking a look at the code.

**Comment by beorn7** [comment]
That's a lot of prepare time. Perhaps most of the query overhead of isolation is happening here? It's not a lot relative to total query time.

The benchmark seems to trigger quite serious peaks in watermark differences. That's not what I saw in production servers under normal query load. I guess that makes the isolation state snapshotting quite expensive…

Anyday, definitely worth some investigation.

**Comment by beorn7** [comment]
I did some profiling.

The additional time spent in prepare seems to be burned mostly in https://github.com/prometheus/prometheus/blob/bf0055536d8463a27725ce9a35ccba55e97d6e5b/tsdb/head.go#L1404 . The code in there has indeed changed a lot between 2.16.0 and now. [This whole section](https://github.com/prometheus/prometheus/blob/bf0055536d8463a27725ce9a35ccba55e97d6e5b/tsdb/head.go#L1408-L1433) is new, replacing a single line we had previously.

This is coming from https://github.com/prometheus/prometheus/pull/6676 , which is actually an optimization.

From my (limited) understanding, I'd say that what we see here is not additional time spent, but just time moved from one phase of the query evaluation to another phase of the query evaluation, i.e. the selection of series within the time range is now done during the prepare phase.

In any case, I think this is likely not related to isolation.

**Comment by roidelapluie** [comment]
/prombench cancel

We will have another long run with the rc.1.

**Pull request merged**